core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java

```diff
@@ -0,0 +1,605 @@
+        protected int fillOffset;
+        protected long fromTimestamp;
+        protected boolean gapFilling;
+        protected long maxTimestamp;
+        protected long minTimestamp;
+        protected long nextBucket;
+        protected BitSet presentRecords;
+        protected int timestampIndex;
+        protected TimestampSampler timestampSampler;
+        protected long toTimestamp;
+        protected ObjList<Function> values;
+        private int rangeBound;
+
+        @Override
+        public void close() {
+            baseCursor = Misc.free(baseCursor);
+        }
+
+        @Override
+        public Record getRecord() {
+            return fillingRecord;
+        }
+
+        @Override
+        public SymbolTable getSymbolTable(int columnIndex) {
+            return baseCursor.getSymbolTable(columnIndex);
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (baseRecord == null) {
+                baseRecord = baseCursor.getRecord();
+            }
+
+            if (baseCursor.hasNext()) {
+                final long timestamp = baseRecord.getTimestamp(timestampIndex);
+                final int bucketIndex = timestampSampler.bucketIndex(timestamp);
+                presentRecords.set(bucketIndex);
+
+                if (timestamp < minTimestamp) {
+                    minTimestamp = timestamp;
+                }
+
+                if (timestamp > maxTimestamp) {
+                    maxTimestamp = timestamp;
+                }
+
+                assert presentRecords.capacity() > bucketIndex;
+
+                return true;
+
+            } else {
+                // otherwise, we need to use bitset to fill
+                if (!gapFilling) {
+                    gapFilling = true;
+
+                    // we need to get up to date, since the sampler started at unix epoch, not min timestamp
+                    if (rangeBound == RANGE_UPPER_BOUND || rangeBound == RANGE_UNBOUNDED) {
+                        fillOffset = timestampSampler.bucketIndex(minTimestamp);
+                        nextBucket = minTimestamp;
+                    }
+
+                    while (recordWasPresent()) {
+                        moveToNextBucket();
+                    }
+
+                    return notAtEndOfBitset();
+                }
+
+                do {
+                    moveToNextBucket();
+                }
+                while (recordWasPresent());
+
+                return notAtEndOfBitset();
+            }
+        }
+
+        @Override
+        public long size() {
+            return -1;
+        }
+
+        @Override
+        public void toTop() {
+            baseCursor.toTop();
+            gapFilling = false;
+            baseRecord = null;
+            fillOffset = 0;
+        }
+
+
+        private Function getFillFunction(int col) {
+            if (col == timestampIndex) {
+                return TimestampConstant.newInstance(nextBucket);
```

Comment:
Allocating a new object on each timestamp access on each bucket will create some litter. Let's avoid it.